### ISSUE TITLE
Read optional properties without check

### ISSUE THREAD

**Issue by NN---** [palantir/tslint#326]
Is it possible to have rule that warns for this code:

``` typescript
interface A { x?: number; }

function f(a :A): number {
 return a.x; // TSLint warning may return undefined.
}
```

And doesn't warn here:

``` typescript
interface A { x?: number; }

function f(a :A): number {
 if (!a.x) 
// or if(a.x !== undefined)
// or if(typeof a.x !== "undefined")
 return a.x; // TSLint warning may return undefined.
else
 return -1;
}
```

**Comment by adidahiya** [comment]
+1

**Comment by dallonf** [comment]
It would be even more helpful if this rule could detect situations like this:

```ts
interface Person {
  address?: {
    line1: string
  }
}

let person = getPerson();
console.log(person.address.line1); // `person.address` might be null or undefined here
```

but would allow:

```ts
if (person.address) {
  console.log(person.address.line1);
}
// or
console.log(person.address && person.address.line1);
```

and, of course, would allow the access of members of required fieds, no questions asked.

I find that these sorts of TypeErrors are still very common even in TypeScript, and it would be great to statically detect them!

**Comment by myitcv** [comment]
```typescript
interface A { x?: number; }

function f(a :A): number {
    return a.x; // TSLint warning may return undefined.
}
```

This is, I think, impossible to achieve sensibly because [Undefined is a sub type of all types](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#327-the-undefined-type). 

```typescript
let a: A = undefined; // this is valid
let n: number = undefined; // as is this
```

Hence you would also have to check that the argument `a` is not undefined... as well as then checking that `a.x` is not undefined.... 

I think you're after the Go equivalent of `struct` and [non-nullable types](https://github.com/Microsoft/TypeScript/issues/185)

**Comment by NN---** [comment]
@myitcv Correct, 'a' has to be checked for undefined too.

```typescript
if (a !== undefined && a.x !== undefined) {
 return a.x; // a.x is not undefined
}
```

**Comment by myitcv** [comment]
@NN--- so under your proposal you would have to guard the usage over every single variable, correct?

**Comment by adidahiya** [comment]
looks like @myitcv is right since we don't have nullable types in TS -- so I think a rule more along the lines of @dallonf's suggestion is more feasible where we lint for nested access on properties that are explicitly optional.

**Comment by dallonf** [comment]
Unfortunately, it seems that "required" vs "optional" - across the board - doesn't mean what we would think and has nothing to do with whether the value is null or not - the only thing that "required" means is that it must be specified. To extend my previous example:

```ts
interface Person {
  address?: {
    line1: string
  }
}

let person : Person = {
  address: undefined
};
```

is perfectly valid according to the compiler. That said, a rule that checks for unprotected access could still be useful - I seem to remember Android Studio running a lint rule like this by default in Java, which doesn't have non-nullable types either.

**Comment by JKillian** [comment]
@dallonf Do you think it's common for people to write TS code where every method call or field access is guarded by a check for null/undefined? In theory I agree with you, but I'm wondering in practice if this is ever what really happens and if a rule that enforces this would ever really be used.

It seems that there are often implicit contracts that aren't enforced explicitly by code:

```typescript
function sum(x:number, y:number) {
  // can reasonable assume that x and y are defined
  return x + y;
}
```

would we really want a rule that forced you to do everywhere:

```typescript
function sum(x:number, y:number) {
  // can reasonable assume that x and y are defined
  if(x != null && y != null) {
    return x + y;
  } else {
    theWorldIsEnding();
  }
}
```

**Comment by dallonf** [comment]
@JKillian IMO it only makes sense to do this for member access (accessing `x` is fine, but `x.y` should be protected), but you're right in that even then it would throw all over the place. Without language support for non-nullable types, it's hard to make this work sanely.

**Comment by NN---** [comment]
Good comments.
I agree that function which accepts 'number' by required and not by optional parameter doesn't expect it to be undefined.
Definitely non-nullable (or non-undefined) types is the better solution for this problem.

I see the only possibility for now is to analyze the caller.
In most cases passing 'undefined' or a variable initialized explicitly with 'undefined' to a function receiving a required argument is an error and not an intended behavior.
Perhaps this case is already handled by TSLint ?

Thanks.

**Comment by andy-hanson** [comment]
We have `--strictNullChecks` in TypeScript now, so I don't think there's any need for this.
